wave_optimization/tests/factor_graph/factor_graph_test.cpp

```diff
@@ -0,0 +1,39 @@
+#include "wave/wave_test.hpp"
+#include "wave/optimization/factor_graph/factor_graph.hpp"
+
+
+namespace wave {
+
+TEST(FactorGraph, addUnaryFactor) {
+    FactorGraph graph;
+    graph.addUnaryFactor<PoseVar, double>(1, 2.0);
```

Comment:
The templates are still there, and are not really the issue.

#### What interface to use?

The change I am suggesting is an interface change from

```cpp
void addUnaryFactor(int, double);
```
to something like:
```cpp
void addUnaryFactor(PoseVar, UnaryFactor);
```

Which IMO is considerably more clear. The above interface decision does not involve templates.

#### But what about the templates?

Of course, the above signatures are what the _template function_ looks like after is generated from a _function template_ that looks like this:
```cpp
template<typename V, typename T>
void addUnaryFactor(int, const &T);
```
or
```cpp
template<typename V, typename T>
void addUnaryFactor(V, T);
```

#### How do templates affect the interface?

Since we _are_ using templates, there is a further difference between the two options: in the first, the compiler cannot deduce the template parameter `V` from a call to the function. That is why the call must [explicitly specify which version of the template you want](https://isocpp.org/wiki/faq/templates#fn-templates-explicit-calls):

```cpp
addUnaryFactor<PoseVar, double>(1, 2.0);
```
vs.
```cpp
addFactor(PoseVar{1}, UnaryFactor{2.0});
```

IMO the second option is again more clear.

Note the second call _could_ be written as 
`addFactor<PoseVar, UnaryFactor>(PoseVar{1}, UnaryFactor{2.0});`
but it is not necessary.

###Template resources
- [C++ FAQ: Templates](https://isocpp.org/wiki/faq/templates#overview-templates)
- [Where and why do I have to put the “template” and “typename” keywords?](http://stackoverflow.com/q/610245/431033)
- _Effective C++_, Chapter 7: Templates and Generic Programming